templates/node/events/index.js

```diff
@@ -13,37 +15,27 @@ const { CloudEvent, HTTP } = require('cloudevents');
  *  -H'Ce-specversion: 1.0' \
  *  http://localhost:8080
  *
- * The event data is extracted from the incoming event and provided as the first
- * parameter to the function. The CloudEvent object itself may be accessed via the
- * context parameter, For example:
- *
- * const incomingEvent = context.cloudevent;
+ * For convenience, the runtime will extract the event from the
+ * CloudEvent envelope and pass it in the second argument to the
+ * function. The entire envelope, along with the HTTP request
+ * metadata, is available from the first argument, 'context'.
  *
  * @param {Context} context the invocation context
- * @param {Object} user the CloudEvent data. If the data content type is application/json
- * this will be converted to an Object via JSON.parse()
+ * @param {Object} event the CloudEvent data
  */
-function verifyUser(context, user) {
-  if (!context.cloudevent) {
-    return {
-      message: 'No cloud event received'
-    };
-  }
+function handle(context, event) {
```

Comment:
I prefer `data` or `eventData` as @zroubalik mentioned.

Reply:
Why?

I know it's just a variable name, but it's obviously important to both of us, so I'm going to reiterate the reasons I gave in slack here:

- The name `data` is not descriptive enough. It's analogous to other poor variable names such as `x` or `object` -- it conveys no domain meaning whatsoever.
- The name `eventData` is too descriptive. It risks confusion with the CloudEvent envelope's `data` field, which is an encoded string. The CloudEvent envelope is an implementation detail that I'd prefer we not expose in the function signature.
- The name `event` is just right. It represents the domain event reconstituted from the CloudEvent envelope according to its `data`, `datacontenttype`, and `dataschema` fields.
- The name `event` is as specific as we can get without knowing which domain is triggering the function. This PR is replacing a function that assumed the domain was user verification. And its second parameter was logically named `user`. It wasn't called `userData` because of course it has data. Otherwise, it wouldn't be a parameter at all!

I genuinely cannot fathom why you guys feel so strong about this. My only guess is that you feel your users will be expecting to work directly with CloudEvent envelopes. And that when they see the parameter name, `event`, they will expect that it's the envelope. If that's true, then we should be passing the envelope there, not the event reconstituted from its data. Personally, I'd rather have the latter, which is why I stand by the name.

But hey, I like to think I'm reasonable -- sell me!

Reply:
Don't know what are anybody else expectations, but when I see `event` I expect it's the whole CloudEvent (including id,source,type etc.).

Reply:
I am not sure if envelope is good word. Bill once compared CloudEvent to  SOAP envelope, but I don't recall the term from CloudEvent spec.

Reply:
btw `data` is optional field for CloudEvent it might be nil

Reply:
but I totally agree that just `data` doesn't tell much

Reply:
> I genuinely cannot fathom why you guys feel so strongly about this. My only guess is that you feel your users will be expecting to work directly with CloudEvent envelopes. And that when they see the parameter name, event, they will expect that it's the envelope. If that's true, then we should be passing the envelope there, not the event reconstituted from its data.

Well IIRC in Go we are working directly with whole CloudEvent, user has to extract details (data) by themself.

Reply:
I am not helping, am I.

Reply:
I think we are mostly coming at this with an awareness of the `CloudEvent` specification and that is influencing how we approach naming. I went back to the spec to see if it offered any illumination for us, and this is what I found under [Terminology](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#terminology).

> An "event" is a data record expressing an occurrence and its context.
> An "occurrence" is the capture of a statement of fact during the operation of a software system. 

The parameter we are discussing is not, in the CloudEvent specification's definition, the event. What is being provided in that parameter is the "statement of fact", expressed as a JavaScript object, extrapolated from the event's representation of the "fact" in some encoded form, typically JSON in the `data` field. While I agree that `data` is not in any way descriptive, I would argue that `event` is barely more so, and in this case I think it would be misleading.

I can understand the concern that adhering to this concept of an event leaks implementation details into the semantics of our API. Frankly, I think that's OK, as the CloudEvent is integral to what we are offering, is a proposed standard, and is the lingua franca of Knative Serverless components. In naming this parameter `event` I think we would be giving the impression that it is, in fact, the CloudEvent. It would lead me to believe that when I call `event.source` to determine where this "statement of fact" originated, I should get a value conforming to the CloudEvent spec's definition of `CloudEvent#source`. But in fact, I will get `undefined`.

Reply:
Helping as much as I am! :)

I don't think we should necessarily strive for cross-language consistency. We shouldn't impose conventions in one language due to limitations in another. I doubt any nodejs developers will care about the quarkus templates.

Regarding the term "envelope", I heard @nainaz and @slinkydeveloper use it, and I like it as a way to distinguish between a conceptual _event_ and its digital _representation_. For me, *envelope* == *CloudEvent*. For you guys building the stuff, it seems like *event* == *CloudEvent*. Not sure about your users (other than myself).

Reply:
In the world of CloudEvents, `event == CloudEvent` based on my understanding of this sentence in the spec.

> An "event" is a data record expressing an occurrence and its context.

The "context" in this sense is all of the event metadata, e.g. `source`, `type`, `id` and the "occurrence" is the data record itself, i.e. `event.data`. 

Providing `event.data` as the second parameter was done as a convenience, assuming that the function developer is primarily interested in the data that was generated by the "occurrence". But maybe there is equal interest in the other event properties as well. For example, `source` and `type` may be used to filter events or perform conditional actions in the function. If that's the case, I would be in favor of an `event` parameter that corresponds to the actual CloudEvent.

Reply:
I'm +1 on the 2nd parameter being an actual CloudEvent if that's the lingua franca of most users. I'm not a "CloudEvent guy" so I apologize for all the noise around this.

I'm assuming for a 2-argument handler, the js-runtime will pass the CE in the second parameter and it will *NOT* be in the context.

And for a 1-argument handler, only the context will be passed, and it will include the CE as it currently does.

Correct? Want me to create an issue in the js-runtime for that?

Reply:
Agree with Lance and Matej, if I read `event` I expect a whole `CloudEvent`, but I might be biased. And because we are using Knative Eventing and users (fn developers) will need to know at least some basic stuff about KEventing and it's concepts, in order to write the functions properly, so therefore I think that a term `CloudEvent` would be at least somehow familiar for them, so eventually they might get the same impression as I do. ie. expect a `CloudEvent` in `event` parameter.

Not sure if changing the 2nd parameter to be a full CE is somehow useful, I suppose that parameter was introduced so users can easily get the actual data 🤷‍♂️

Reply:
When implementing a Handler with Naina (having no familiarity with the function signatures of the language beforehand), we were both thrown off by the second parameter not being a CloudEvent.  It seemed an obvious assumption that the _event_ would be the argument.  Or, at least this would be a supported default.  I would certainly advocate for that change, while at the same time completely agreeing with Jim that we should not force consistency between languages; instead working to their strengths.